### PR TITLE
python3Packages.torch-bin, python3Packages.torchvision-bin, python3Packages.torchaudio-bin: update cuda 11.6 -> 11.7

### PR DESCRIPTION
**Pull request by GaetanLepage**
###### Description of changes

Updated the cuda version that torch-bin, torchvision-bin and torchaudio-bin are built on from 11.6 to 11.7.
This is the cuda version currrently shipped in NixOS and it is the one that this version of pytorch is officially supporting.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

###### Testing

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages built:</summary>
  <ul>
    <li>python310Packages.torch-bin</li>
    <li>python310Packages.torchaudio-bin</li>
    <li>python310Packages.torchvision-bin</li>
    <li>python310Packages.trainer</li>
    <li>tts</li>
  </ul>
</details>

@junjihashimoto

### CODE REVIEW

**Comment by GaetanLepage** [comment]
I want to add that this PR unbreaks the build process of torch-bin.
(currently broken on master)

**Review by mweinelt — CHANGES_REQUESTED** [review]
One commit per attribute please.

**Comment by GaetanLepage** [comment]
> One commit per attribute please.

Done :)

**Comment by SomeoneSerge** [comment]
Result of `nixpkgs-review pr 215091` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python310Packages.torchaudio-bin</li>
    <li>python310Packages.torchvision-bin</li>
    <li>python310Packages.trainer</li>
    <li>tts</li>
  </ul>
</details>

**Review by wegank — APPROVED** [review]

File: pkgs/development/python-modules/torchvision/prefetch.sh [hunk 1]
```diff
@@ -5,7 +5,7 @@ set -eou pipefail
 
 version=$1
 
-linux_cuda_version="cu116"
```
Comment (SomeoneSerge) [inline pkgs/development/python-modules/torchvision/prefetch.sh:8]: I haven't thoroughly looked at torch `-bin` packages before. Shouldn't this be synchronized with `cudaPackages.cudaVersion`? I also don't see any `autoPatchelfHook`s in `bin.nix`, so maybe I just misunderstand how this derivation works

Comment (GaetanLepage) [reply pkgs/development/python-modules/torchvision/prefetch.sh:8]: From what I have seen, the `-bin` derivations only download the appropriate binaries from torch website.
Those binaries come bundled with their own version of cuda, which does not need to be the same as the one on the system.
Hence, it should not be linked to the `cudaPackages.cudaVersion` and we don't need to use `autoPatchelfHook`.

As for which version of cuda to choose (the `cu117` in `prefetch.sh`), it is the one that torch ships by default with the latest version of each package.
https://pytorch.org/get-started/locally
i.e., the cuda version that yields to the canonical install command:
```sh
pip install torch torchvision torchaudio
```

**Review by junjihashimoto — APPROVED** [review]
Thx!

**Review by SomeoneSerge — APPROVED** [review]

**Review by samuela — APPROVED** [review]

**Pull request merged**